src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts

```diff
@@ -36,13 +40,22 @@ export class CrushmapComponent implements OnInit {
   metadataTitle: string;
   metadataKeyMap: { [key: number]: any } = {};
 
-  constructor(private healthService: HealthService) {}
+  constructor(private healthService: HealthService, private timerService: TimerService) {}
 
   ngOnInit() {
     this.healthService.getFullHealth().subscribe((data: any) => {
       this.loadingIndicator = false;
       this.nodes = this.abstractTreeData(data);
     });
+    this.sub = this.timerService
+      .get(() => this.healthService.getFullHealth(), 5000)
+      .subscribe((data: any) => {
+        this.nodes = this.abstractTreeData(data);
+      });
+  }
+
+  ngOnDestroy() {
+    this.sub.unsubscribe();
```

Comment:
You can simplify this even further. A nice way not to having to take care of subscribing/unsubscribing to observables is moving them to the HTML template (as they only serve as inputs for other components, right?) and use the [`async` pipe](https://malcoded.com/posts/angular-async-pipe/).

So we move this timer observable to the template without subscribing to it in the TS file:
```suggestion
    this.data$ = this.timerService
      .get(() => this.healthService.getFullHealth(), 5000);
  }
```

However the health endpoint is an amalgamated one (hence the `abstractTreeData` function... why not having a separate crushmap one?), so every time this observable polls the health endpoint, it will return a lot of changing data (which will trigger Angular change detection process), but we only want what's in `data.osd_map.tree.nodes`... And that will barely change. What about creating an observable that only picks those changes?

```ts
    this.node$ = this.timerService
      .get(() => this.healthService.getFullHealth(), 5000)
      .pipe(
           pluck('osd_map', 'tree', 'nodes'),
           distinctUntilChanged(),  // this avoids emitting new events when no actual changes
      ); // this is the 'nodes' data
  }
```

Additionally we need to simplify the `abstractTreeData` (IMHO this kind of data transforming method should be in the back-end, not in the front-end):
```diff
-  private abstractTreeData(data: any): any[] {                                                          
+  private abstractTreeData(nodes: any): any[] {                                                          
-    const nodes = data.osd_map.tree.nodes || [];                                                        
    const treeNodeMap: { [key: number]: any } = {};                                                     
                                   
```
And then, in the HTML template, the node data is ready to be used:

```html
<i *ngIf="node$ | async as node">
  ...
  <tree-root #tree 
    [nodes]="node | pipeFunction: abstractTreeData" 
```

Reply:
We could also invoke the abstractTreeData with `map()` operator, and in the HTML no having any pipeFunction at all. Definitely, I don't get any good reason to have data processing in the front-end...

Reply:
As per the discussion in today's standup the improvement of avoiding data processing in frontend will be covered in a follow-up PR and this is the tracker for the same https://tracker.ceph.com/issues/49236

Reply:
Thanks @avanthakkar !